Navbar: Add UI for Repo Filter

danielbeardsley opened the pull request:
Add a UI to easily show / hide pulls from individual repos.

Meets all the requirements from the issue:

* UI component 
   * multi-select list of repos
   * Selecting *none* of the options reset's to selecting all
   * updates immediately
   * populates list from the open pulls
   * Show the pull count next to each repo name
* Repos
   * list is parsed from the url
   * url is updated when changing the UI component
   * allow repos that don't have pulls to remain in the list if they are parsed from the URL

Closes #268

danielbeardsley commented:
![image](https://user-images.githubusercontent.com/26855/161862549-f12f1387-f126-48a7-aa14-f8d67bf4b1d0.png)

deltuh-vee commented:
dev_block ✊ 
UI component
- [x] multi-select list of repos
- [x] Selecting none of the options reset's to selecting all
- [x] updates immediately
- [ ] populates list from the open pulls
![CleanShot 2022-04-06 at 09 47 46](https://user-images.githubusercontent.com/5289715/162026210-7c873c45-ce03-4dce-9764-6ce8a14559b7.png)
What pull is in this Repo?
- [x] Show the pull count next to each repo name

Repos
- [x] list is parsed from the url
- [x] url is updated when changing the UI component
- [x] allow repos that don't have pulls to remain in the list if they are parsed from the UI

danielbeardsley commented:
For clarity:

Right now, the counts next to the repos include *all open pulls*. If we want that to be different, we'll have to be specific about what the counts should be. Thoughts @deltuh-vee ?

deltuh-vee commented:
@danielbeardsley I personally don't find the counts next to the repo to be very helpful, so I'm actually fine with them not being shown.

danielbeardsley commented:
OK. I think I'll un_dev_block :+1:  then cause I find the counts useful and I can't really think of better values than the ones we have.

deltuh-vee commented:
QA 🎬 
Repo list is populated from open pulls.

andyg0808 commented on frontend/src/use-url-state.ts:
```diff
@@ -18,18 +18,15 @@ type HistoryState = Record<string, string>;
+   });
+
+   // Watch for url state changes (back button) only once per component.
+   useEffect(() => {
+      return watchForPopstate(paramName, setState, paramDefault);
```
Pedantic, but this could be `useEffect(() => watchForPopstate(paramName, setState, paramDefault));` now if you like.

danielbeardsley replied:
Fixed!

andyg0808 commented on frontend/src/repo-menu.tsx:
```diff
@@ -0,0 +1,59 @@
+      }
+
+      // Update the url
+      setSelectedRepos(newSelectedRepos);
+
+      // Update the pull filter
+      const selectedReposSet = new Set(newSelectedRepos);
+      setPullFilter('repo', selectedReposSet.size === 0 ? null : (pull) =>
+         selectedReposSet.has(pull.getRepoName())
+      );
```
I'm betting the fact that there are two places with similar state like this is going to bite us eventually. We may want to move toward allowing the pull filters to be based on the URL state or something down the road.

Redux may be worth considering for some of this, because it provides a global location to store state which can be accessed from multiple places.

danielbeardsley replied:
My first reaction was "Yes, url is the state, that's how it was done". So I looked more closely at the code and you are totally right, that's *not* how it's currently done and there's a bug: history navigation is broken!

Fixing!

danielbeardsley replied:
Fixed.

andyg0808 commented on frontend/src/repo-menu.tsx:
```diff
@@ -49,11 +53,18 @@ export function RepoMenu() {
 }
+
+function getRepoToPullCount(pulls: Pull[]): RepoCounts {
+   return pulls.reduce((counts, pull) => {
+      counts.set(pull.getRepoName(), (counts.get(pull.getRepoName()) || 0) + 1);
+      return counts;
+   }, new Map() as RepoCounts);
```
Might be worth doing this as a [`countBy`](https://lodash.com/docs/4.17.15#countBy), since it makes it clear exactly what we're doing:
```
return _.countBy(pulls, pull => pull.getRepoName());
```
(you'll need to switch the return type to `Record<string, number>` or something like that, I think).

danielbeardsley replied:
Good idea, fixed!

andyg0808 commented on frontend/src/repo-menu.tsx:
```diff
@@ -0,0 +1,80 @@
+   const urlRepos = useConst(() => new Set(selectedRepos));
+   const setPullFilter = useSetFilter();
+
+   // May include repos frmo the url for which there are no pulls
```
```suggestion
   // May include repos from the url for which there are no pulls
```

andyg0808 commented on frontend/src/repo-menu.tsx:
```diff
@@ -0,0 +1,80 @@
+   const allRepos = useMemo(() => {
+      // All repos of open pulls
+      const pullRepos = new Set(pulls.map((pull) => pull.getRepoName()));
+      return [...new Set([...pullRepos, ...urlRepos])]
```
Depending, might also be worth sorting this array:
```suggestion
      return _.sortBy([...new Set([...pullRepos, ...urlRepos])])
```

or (happens to work because [`sort()` returns the array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort)):
```suggestion
      return [...new Set([...pullRepos, ...urlRepos])].sort();
```

danielbeardsley replied:
Fixed in other pull.

andyg0808 commented:
CR :goat: through b8807e8
deploy_block :+1: on optional comment.

danielbeardsley commented:
Great CR @andyg0808 !!

Since the author-filter pull refactors this code significantly, I've addressed your concerns there. See the recent commits at #285 

I think I'm going to close this pull and just let the other one do the master merge.
